domain/application/service/storage.go

```diff
@@ -0,0 +1,168 @@
+// - [github.com/juju/juju/domain/storage/errors.StorageNotFound] when the storage doesn't exist.
+// - [github.com/juju/juju/domain/application/errors.UnitNotFound]: when the unit does not exist.
+// - [github.com/juju/juju/domain/application/errors.WrongStorageOwnerError]: when the storage owner does not match the unit.
+// - [github.com/juju/juju/domain/application/errors.UnitNotAlive]: when the unit is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNotAlive]: when the storage is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AttachStorage(ctx context.Context, storageID corestorage.ID, unitName coreunit.Name) error {
+	if err := unitName.Validate(); err != nil {
+		return errors.Capture(err)
+	}
+	if err := storageID.Validate(); err != nil {
+		return errors.Capture(err)
+	}
+	unitUUID, err := s.st.GetUnitUUIDByName(ctx, unitName)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	storageUUID, err := s.st.GetStorageUUIDByID(ctx, storageID)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	err = s.st.AttachStorage(ctx, storageUUID, unitUUID)
+	if errors.Is(err, applicationerrors.StorageAttachmentAlreadyExists) {
+		return nil
+	}
+	return err
+}
+
+// AddStorageForUnit adds storage instances to given unit.
+// Missing storage constraints are populated based on model defaults.
+// The following error types can be expected:
+// - [github.com/juju/juju/core/unit.InvalidUnitName]: when the unit name is not valid.
+// - [github.com/juju/juju/core/storage.InvalidStorageName]: when the storage name is not valid.
+// - [github.com/juju/juju/domain/storage/errors.StorageNotFound] when the storage doesn't exist.
+// - [github.com/juju/juju/domain/application/errors.UnitNotFound]: when the unit does not exist.
+// - [github.com/juju/juju/domain/application/errors.UnitNotAlive]: when the unit is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNotAlive]: when the storage is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AddStorageForUnit(
+	ctx context.Context, storageName corestorage.Name, unitName coreunit.Name, stor storage.Directive,
```

Comment:
is `stor` a typo?

Reply:
no, i just didn't want to use "storage" and "store" implies something else. 
Renamed to "directive".